usr/sbin/p11sak/p11sak.c

```diff
@@ -112,53 +112,44 @@ static void load_pkcs11lib(void)
     atexit(unload_pkcs11lib);
 }
 
-static CK_RV get_pin_prompt(char **pin, size_t *pinlen)
+static char *get_pin_prompt(void)
 {
     struct termios old, new;
     int nread;
-    char *user_input = NULL;
-    size_t buflen = 0;
-    CK_RV rc = 0;
+    size_t pinlen = 0;
+    char *pin = NULL;
 
     printf("Please enter user PIN: ");
 
-    /* turn echoing off */
+    /* turn echo off */
     if (tcgetattr(fileno(stdin), &old) != 0)
-        return -1;
+        goto out;
 
     new = old;
     new.c_lflag &= ~ECHO;
+
     if (tcsetattr(fileno(stdin), TCSAFLUSH, &new) != 0)
-        return -1;
+        goto out;
 
-    /* read the pin
-     * Note: getline will allocate memory for user_input. free it when done.
-     */
-    nread = getline(&user_input, &buflen, stdin);
-    if (nread == -1) {
-        rc = -1;
-        goto done;
-    }
+    /* note: getline will allocate memory for pin. */
+    nread = getline(&pin, &pinlen, stdin);
+    if (nread == -1)
+        goto out;
```

Comment:
I will move the getline() error handling below the re-enable of the console echo. Otherwise, the console echo stays disabled on getline() failures and the caller is required to reset the console.